### ISSUE TITLE
Asynchronously save queue items

### ISSUE THREAD

**Issue by fredrikekelund** [cgiffard/node-simplecrawler#257]
Building on #253, I'd like to look into how we mutate queue items. Currently it's all done synchronously, and I'd like to save them asynchronously after they're mutated, perhaps via an upsert method on the queue that takes two arguments, `id` and a `queueItem`. With this, we'd truly enable different queue backing stores, and even make it really easy to run the crawler in a distributed fashion.

**Comment by cgiffard** [comment]
Good point — and something I'd previously missed. If we can make this work without complicating the queue interface, we're in a good position.

**Comment by fredrikekelund** [comment]
I think so too, it struck me how easy it would be to run simplecrawler in a cluster if we had this.

And I don't think this would necessarily complicate the queue interface, but I do think Promises could provide a real benefit here. What are your thoughts on bringing in a dependency like Bluebird?

**Comment by cgiffard** [comment]
Yeah, the intention behind a DB-backed queue was actually so you could run simplecrawler in a cluster. My preference with dependencies is actually to keep things really lean, and let consumers of the API wrap their code in promises if they want it — keeping callbacks underneath.

But, I think it would be easier to make the decision if we first identified all the code which is mutating the queue, and how easy it would be to write to a given pattern.

**Comment by fredrikekelund** [comment]
Alright, I can sympathise with that. I'll see if I have time this weekend to bang out an initial implementation!

**Comment by fredrikekelund** [comment]
Building this comment https://github.com/cgiffard/node-simplecrawler/pull/256#discussion_r63279012, I think that `crawler.queue.upsert` ought to take 2 arguments, a URL that works as an id, and an object that describes the changes to be made to the queueItem that represents that URL. In the default queue implementation, this basically means we'll use `Object.assign(originalQueueItem, updatedQueueItem)`

**Comment by fredrikekelund** [comment]
I'll have a look at this tonight! This and #258 should be the remaining two things that need to land before 1.0.0 (besides writing some good release notes along with an upgrade path)

**Comment by jtlapp** [comment]
I'm looking for a crawler that I can back with a robust queue on PostgreSQL. Looking over the queue interface just now, I'm wondering if it might require revision to support the necessary atomicity of queue transactions.

In particular, it seems to me that oldestUnfetchedItem() is the only query for which the crawler could trust results. Anything else is asking the queue for its current state, which might change before the state is subsequently acted on.

Or are methods such as FetchQueue::get(), FetchQueue::getLength(), and anything that does an analysis of the queue used only for approximate informational purposes?

I would implement a queue that locked a job on oldestUnfetchedItem() until the goods are delivered by callback. Then I would delete the job and commit the fetch (after processing the data). Any analysis would have to be done on a job archive. Would it be okay for methods such as get() and getLength() to only access an archive of completed jobs? For which the job may not be available? Sorry for not investigating the code at length first; I'd like to have a little confidence before investing the time. Thanks for your help!

**Comment by jtlapp** [comment]
Alternatively, I could leave the job in the queue upon completion, marking it as complete and completing the transaction begun at fetch. I would then have a separate process grab completed jobs from the queue for further processing, removing them from the queue as processing completes. Now we have multiple processes hitting the same queue, as envisioned in the OP. Seems like I face the same problems, though.

**Comment by jtlapp** [comment]
(Sorry, I tend to be overly communicative.) If my application seems within scope of this module, maybe I could be simultaneously working on a PostgreSQL queue, and together we could hammer out a suitable interface. I'm leaning toward using https://github.com/timgit/pg-boss.

**Comment by cgiffard** [comment]
(I'm currently at work so have only glanced superficially at your comment — apologies for that.) I think unless there's something I'm missing, the atomicity and transaction integrity could absolutely be handled by the queue — it could lock certain async methods until others had returned.

I think only `getOldestUnfetchedItem` is really the only retrieval operation which will require locking, at least for the operation of the crawler itself, but you might find that clients consuming the queue itself want more integrity for their more 'informational' calls (printing stats, etc.) I think how much you care about the transactional integrity of those things is entirely up to you!

**Comment by jtlapp** [comment]
I suspect there might be a more modular way to separate concerns, so that queue-smart people can write queue wrappers, and application writers can just pick a queue and separately write the application logic. I fear that this interface may require wrapping the queue in an application layer.

I think we want to be able to say that a queue must pass the queue test suite in order to assure that it works with the crawler. Anything that isn't strictly necessary for queuing might be hung off of a different component, which could be either application-specific or generic.

I'll have a closer look. But the queue interface does seem to impose unnecessary requirements.

**Comment by cgiffard** [comment]
@jtlapp Which requirements do you see as onerous or unnecessary? If we can come up with some simple ways to streamline the queue, we might be able to make it simpler for implementors.

**Comment by cgiffard** [comment]
(As an aside, I think any queue streamlining, or the addition of locking to the crawler or queue interface should be a separate task and should not block any work on #301.)

**Comment by jtlapp** [comment]
I'll have to look closely at the code tomorrow. I should make sure that my intended use is within scope of this module before proceeding. Here's what I envision:

- The web crawler will run in a background process, either working through a job queue or waiting to wake up via IPC.
- A main process will listen to HTTP and ask the web crawler process to load HTML pages.
- When an HTML page loads, domain-specific logic will identify images within the web page and queue those for loading. No need to save the HTML page or its job.
- When an image completes loading, I write the image to a file, and I queue the image, along with some metadata, for conversion to a thumbnail. (Actually, I'd like to be writing the image to the filesystem as it loads, to conserve RAM.)
- A third process, likely running on a second computer, will poll this image queue and convert images to thumbnails, storing those thumbnails to disk.

This can all be done in one queue, of course, but some queue implementations are more efficient at smaller sizes, particularly when many of the rows may be locked in transactions. I envisioned reporting images to the third process via a second queue that is not of concern to the web crawler. In particular, I'm not seeing why the web crawler would need to update jobs (write back to the queue) in my application.

Does this fall within scope? I'm not keen on storing data in the DB that never gets used. I'm going to have huge amounts of data to deal with for other reasons.

**Comment by jtlapp** [comment]
I've done a quick assessment of the code from the perspective of the queue. Here are my findings:

(1) Outside the queue itself, `get`, `exists`, `filterItems`, `max`, `min`, and `avg` are currently only used by the queue test suite. These are not required queue methods. I don't think any are absultely needed for testing, either. They may be useful for some applications, but depending on the application, the implementation may or may not have the crawler's job queue underneath.

(2) In order for me to run the crawler as a recoverable background process, the caller will need to be able to get its first job from the queue rather than from `initialURL`.

(3) The crawler calls `countItems()` and `getLength()` when `oldestUnfetchedItem()` fails to return a new job and there are neither any open requests nor any open listeners. If the number of jobs performed equals the queue size, the crawler quits. This code makes these assumptions:

    (a) That every job ever performed will remain in the queue until the crawler is done. No other entity can remove from the queue until the crawler finishes.
    (b) That the thread in which the crawler runs is the only thread adding jobs to the queue. If the crawler reaches the `countItems()` test at the instant another thread is adding a new job to the queue, the crawler may quit before seeing the new job.
  
(4) `freeze()` and `defrost()` clearly only apply to in-memory queues, but they are an important part of the in-memory queue interface.

It seems that there are two potential uses cases of a web crawler that are at odds here. One is that a crawler runs to perform a single, finite task that consists of a set of jobs kept in its own queue. The other is that the crawler runs indefinitely for an unending task of an indeterminate number of jobs on a possibly-shared queue.

It should be possible to write one module that supports both use cases, but that's a design decision. I think the more general use case would use the following architectural partitioning: job queue, client that posts to job queue, web crawler that services job queue, page/file processor that receives completed jobs and may post to job queue, library of nice-to-have page-reading utilities.

**Comment by jtlapp** [comment]
If you want multiple crawlers to be able to service the same queue, I suspect you'll need to support the second use case.

**Comment by fredrikekelund** [comment]
@jtlapp great to have someone interested in implementing a distributed queue! Lots to respond to here, but here are my thoughts:

- the crawler should indeed be able to trust the results of all calls to the queue (I think), since any methods that return statistics act only on fetched queue items, and queue items are only fetched once
- race conditions where one crawler instance completes just before another one adds more items to the queue are of course a risk, but in general, the number of items in the queue should increase at a higher rate than the number of completed requests. When crawling a website, it wouldn't be uncommon to have over 100 items in the queue after just 5 completed requests. The biggest issue ought to be the time before the first request has completed, when no other items are in the queue. We'll have to devise a way to work around that. 
- glancing through the use case you're describing, I think that the crawler part of such an application doesn't have to be all that modular. If you need to scrape web pages for images, then the crawler will download both HTML pages and images by default. Heavy image processing jobs might benefit from being executed in a separate process, but the crawler could perform the fetching from one or multiple instances (provided there's a distributed queue implementation of course). I might be misunderstanding your requirements though! Be happy to discuss further on that case

Maybe we can keep further discussion on this matter to a separate issue if it isn't at odds with the changes introduced here? :)

**Comment by jtlapp** [comment]
I'm not sure. Your response isn't encouraging. The stats methods won't work in my case because the relevant items won't even be in the queue. Race conditions are only something you allow to exist (at all) in applications that don't require reliability. And I'll be taxing resources to such a degree that I must have full control over what URLs are downloaded at all.

**Comment by jtlapp** [comment]
I've done some investigation into what it would take for me to write this from scratch. What I want is very simple. It could even serve as a core for your load-until-done design paradigm, while also allowing you to cluster crawlers on a common queue (after you make a few tweaks). I'm in a hurry for this piece, so I should have something to share soon. I'll do it as a separate module.

**Comment by cgiffard** [comment]
Thanks again for sharing your approach! It's great to hear from people (even criticism) so we can improve the project.

The avoidance of race conditions are obviously important for us, but I must admit the particular method of modularity you're describing doesn't seem like the domain of the /crawler/, or at least this specific module, but you could easily go down that road in your own implementation if you wanted to.

Generally speaking, options for Node IPC that aren't absolutely abhorrent are few and far between, and the domain we're operating in here, synchronising relatively heavy objects relies on a lot of (I would argue, unavoidable) CPU-heavy serialisation which splitting the queue into not one, but two process boundaries will only compound. Perhaps I'm misunderstanding your intention here — apologies if that's the case.

As an additional aside, many larger crawls I've done so far (in the hundreds of millions of resources) have spent most of their time waiting on network, and very little time processing — even when handling and archiving large binary files. I was surprised to learn that you anticipate CPU overheads being significant, and am curious as to whether there's anything specific about your use case (beyond the processing of the images themselves, which is a separate concern) that would lead you to make this assertion.

The intention of FetchQueue is to provide a relatively generic interface for any kind of queue one might want to suit their task at hand — whether that's for archiving a little site to disk or doing a massive distributed crawl. The points you raise about the assumptions of queue length are valid, and warrant separate discussion. If we can improve the queue interface to satisfy your use case, this is absolutely something we want to do!

That said, if methods like stats generation are not useful to you — you don't have to implement them. These are methods provided by the in-memory queue bundled with the crawler, and as you rightly point out, they're not used by the crawler itself.

Conceptually speaking, we could split these areas of duty down the middle, and say that the in-memory queue implements two interfaces — the FetchQueue the crawler needs, and a set of handy tools for client implementors (catchy names pending, of course!) This would enable you to implement just the things that matter for your case, and do so in a way which is supported by the crawler (i.e. we won't just start relying on stats generation in the crawler core.) But I believe this should be a change in the API we guarantee and support, rather than involve any kind of mandatory separation.

Furthermore, a queue isn't necessarily a 1:1 mapping to a DB. If it makes sense for you to remove items from the queue to save space in your DB, you can always use alternative methods to track what items you've already fetched, and what the length of the queue is as seen by the crawler — this could be a monotonic value stored in Postgres, or Redis, or read in the pattern of the stars via telescope from the far side of the moon. The back-end implementation of the queue is entirely up to you — and the word 'queue' is not intended to constrain your thinking about how this happens. :)

**Comment by jtlapp** [comment]
You're making more sense to me now. ;-) Under the approach I envision, the "crawler" would be distinct from the HTTP web client. The web client will dutifully perform queue jobs and message any subscribed listeners. The crawler would be a subscriber that also posts new jobs to the queue.

The web client could be configured to poll the queue. Postgres has a NOTIFY facility for pushing messages out to subscribers. IPC would just be to issue a wakeup call, in case the application hasn't heard from it in a while and suspects it might have gone dormant. Polling would be necessary in a distributed configuration.

Some apps won't need a job archive. Mine can't have one, because there is no end to the jobs. In any case, counting lots of rows in a database makes for a time-consuming query. I also don't think it's needed. The application can monitor job completions and inform the crawler that it's done. In your case, the "application" could just be a smart layer around the core.

I was going to use the `request` module https://github.com/request/request. Any reason why you took the challenge of coding to `http` and `https` directly? The `request` module also provides for direct piping to the file system, which I'd make configurable per-job. I don't think I'll need to store HTML, unless my web scraping proves time-consuming.

**Comment by cgiffard** [comment]
This module is *really* old (for a node module, anyway!) — it must be getting toward five or six now. Anyway, when I first wrote it, `request` hadn't hit the scene, or wasn't stable. And by the time it was the given way to do things, we already had a pretty battle-tested implementation. If I had to write a new crawler now, I would probably use `request`. (If I had to write a crawler back in 2010 or 2011 or whenever it was, knowing what I know now, I would've *written* `request`. Heh!)
